File: prototype/rails/src/app/domain/payments/card_decision.rb
module Domain
  module Payments
    # What the card processor said about one number.
    CardDecision = Data.define(:last_four, :decline_reason) do
      def self.approved(last_four)
        new(last_four: last_four, decline_reason: nil)
      end

      def self.declined(last_four, decline_reason)
        new(last_four: last_four, decline_reason: decline_reason)
      end

      def approved?
        decline_reason.nil?
      end
    end
  end
end

File: prototype/node/src/app/core/payments/card-decision.ts
import type { DeclineReason } from './decline-reason.ts'

/** What the card processor said about one number. */
export type CardDecision = {
  lastFour: string
  declineReason: DeclineReason | null
}

export function approvedCard(lastFour: string): CardDecision {
  return { lastFour, declineReason: null }
}

export function declinedCard(lastFour: string, reason: DeclineReason): CardDecision {
  return { lastFour, declineReason: reason }
}

export function isCardApproved(decision: CardDecision): boolean {
  return decision.declineReason === null
}
